src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java

```diff
@@ -0,0 +1,781 @@
+  private final ListImpl list = new ListImpl();
+
+  // Whether the pool is closed
+  private boolean closed;
+
+  // Selectors
+  private BiFunction<PoolWaiter<C>, List<PoolConnection<C>>, PoolConnection<C>> selector;
+  private BiFunction<PoolWaiter<C>, List<PoolConnection<C>>, PoolConnection<C>> fallbackSelector;
+
+  // Connection state
+  private final Slot<C>[] slots;    // The pool connections, this array is not sparse
+  private int size;                 // The number of non null slots
+  private int capacity;             // The pool capacity
+
+  // The queue of waiters
+  private final Waiters<C> waiters;
+
+  SimpleConnectionPool(PoolConnector<C> connector, int[] maxSizes) {
+    this(connector, maxSizes, -1);
+  }
+
+  SimpleConnectionPool(PoolConnector<C> connector, int[] maxSizes, int maxWaiters) {
+
+    int[] capacities = new int[maxSizes.length];
+    int maxCapacity = 1;
+    int numSlots = 0;
+    for (int i = 0;i < maxSizes.length;i++) {
+      int maxSize = maxSizes[i];
+      if (maxSize < 1) {
+        throw new IllegalArgumentException();
+      }
+      maxCapacity *= maxSize;
+      numSlots = Math.max(numSlots, maxSize);
+    }
+    for (int i = 0;i < maxSizes.length;i++) {
+      capacities[i] = maxCapacity / maxSizes[i];
+    }
+
+    this.capacityFactors = capacities;
+    this.connector = connector;
+    this.slots = new Slot[numSlots];
+    this.size = 0;
+    this.maxWaiters = maxWaiters;
+    this.capacity = 0;
+    this.maxCapacity = maxCapacity;
+    this.sync = new CombinerExecutor2<>(this);
+    this.selector = (BiFunction) SAME_CONTEXT_SELECTOR;
+    this.fallbackSelector = (BiFunction) FIRST_AVAILABLE_SELECTOR;
+    this.waiters = new Waiters<>();
+  }
+
+  @Override
+  public ConnectionPool<C> connectionSelector(BiFunction<PoolWaiter<C>, List<PoolConnection<C>>, PoolConnection<C>> selector) {
+    this.selector = selector;
+    return this;
+  }
+
+  private void execute(Executor.Action<SimpleConnectionPool<C>> action) {
+    sync.submit(action);
+  }
+
+  public int size() {
+      return size;
+  }
+
+  public void connect(Slot<C> slot, PoolWaiter<C> waiter) {
+    slot.initiator = waiter;
+    connector.connect(slot.context, slot, ar -> {
+      slot.initiator = null;
+      if (ar.succeeded()) {
+        execute(new ConnectSuccess<>(slot, ar.result(), waiter));
+      } else {
+        execute(new ConnectFailed<>(slot, ar.cause(), waiter));
+      }
+    });
+  }
+
+  private static class ConnectSuccess<C> implements Executor.Action<SimpleConnectionPool<C>> {
+
+    private final Slot<C> slot;
+    private final ConnectResult<C> result;
+    private PoolWaiter<C> waiter;
+
+    private ConnectSuccess(Slot<C> slot, ConnectResult<C> result, PoolWaiter<C> waiter) {
+      this.slot = slot;
+      this.result = result;
+      this.waiter = waiter;
+    }
+
+    @Override
+    public Runnable execute(SimpleConnectionPool<C> pool) {
+
+      int capacity = pool.capacityFactors[(int)result.weight()];
```

Comment:
Shouldn't `result.weight()`be the `kind`?

Reply:
it actually is, it is just that result.weight() needs to be renamed after we remove the current connection pool the `ConnectResult` class belongs to.